### DISCUSSION TITLE
FluidSynth should not reset synth before first MIDI file is played

### DISCUSSION THREAD

**Discussion by mawe42** [FluidSynth/fluidsynth#816] [Ideas]
While investigating #102 I wondered why certain commands I added to a user command file had no effect on the MIDI file that I tried to play. Later I noticed that I stumbled (yet again) over the fact that the player resets the synth before each MIDI song by default, and that adding `-o player.reset-synth=0` solves that problem. And then I remembered that we had quite a few support requests in the past where users wondered why their commands were ignored, and were told to unset the reset-synth option.

That made me wonder: is there a use-case for a synth reset before the *first* song? I can understand why we would need a reset between songs (and the variable name `reset_synth_between_songs` seems to reflect this), but why before the first one?

If there is no real use-case, then I would propose changing that behaviour. I think it would be quite easy, for example by adding a new `is_between_songs` member to `fluid_player_t` that gets set after the first check for `reset_synth_between_songs`:
```diff
@@ -2057,14 +2059,17 @@ fluid_player_playlist_load(fluid_player_t *player, unsigned int msec)
     player->begin_msec = msec;
     player->start_msec = msec;
     player->start_ticks = 0;
     player->cur_ticks = 0;
 
-    if(player->reset_synth_between_songs)
+    /* make sure we only reset the player *between* songs if requested,
+     * not before the first song. */
+    if(player->is_between_songs && player->reset_synth_between_songs)
     {
         fluid_synth_system_reset(player->synth);
     }
+    player->is_between_songs = TRUE;
 
     for(i = 0; i < player->ntracks; i++)
     {
         if(player->track[i] != NULL)
         {

```

**Comment by jjceresa** [comment]
>If there is no real use-case, then I would propose changing that behaviour. I think it would be quite easy, for example by adding a new is_between_songs member to fluid_player_t that gets set after the first check for reset_synth_between_songs.

Instead of adding a new `is_between_songs` member we simply need to move the check for `fluid_synth_system_reset` at the end of the file playing.

**Reply by mawe42** [reply]
That could be easier, not sure. Do we actually have an "end of file" condition in the player? What if somebody used `player_next` before the end of the song?

**Comment by jjceresa** [comment]
> Do we actually have an "end of file" condition in the player?

Here. It works also if somebody used `player_next`, because this command simply does a seek to the end of current song leading to load the next song (if any).
```
@@ -2145,6 +2145,7 @@ fluid_player_callback(void *data, unsigned int msec)
         {
             FLUID_LOG(FLUID_DBG, "%s: %d: Duration=%.3f sec", __FILE__,
                       __LINE__, (msec - player->begin_msec) / 1000.0);
+            printf("end-of-song\n");
             loadnextfile = 1;
         }
```

**Reply by mawe42** [reply]
Very good idea, thanks! So the technical solution would be fairly simple... the question that remains: is there a valid use-case for a synth reset before the first song? I'm fairly convinced that it is unnecessary for use-cases involving the command-line program. But I'm unsure about API users here, as I have never used the player API in a program before.

**Comment by derselbst** [comment]
Since a clean synth has been created, just before the first MIDI is being played, I see no use-case to reset the synth again. I do consider it more as a bug, as it undoes changes made by the user command file.

**Comment by jjceresa** [comment]
>Since a clean synth has been created, just before the first MIDI is being played, I see no use-case to reset the synth again.

Yes. Also, moving the check for `fluid_synth_system_reset` at the `end of current song playing` ensures that the synth regains its clean state after the last MIDI song will be finished.

**Comment by mawe42** [comment]
Ok, thanks for your feedback. I'll prepare a PR to address this.
